File: archive/2020f/solutions/TheFlockofRams.cpp
#include <bits/stdc++.h>

// #include "testlib.h"

using namespace std;

typedef long long ll;

ll n, m, k;
const ll dr[4] = {0, 0, -1, 1};
const ll dc[4] = {1, -1, 0, 0};

bool valid(ll r, ll c) {
	return r >= 0 && r < n && c >= 0 && c < m;
}

void solve(int argc, char* argv[], int tc = 0) {
	cin >> n >> m >> k;
	
	bool grid[n][m];
	for (ll i = 0; i < n; i++) {
		string s;
		cin >> s;
		for (ll j = 0; j < m; j++) {
			grid[i][j] = (s[j] == '#');
		}
	}
		
	pair<ll, ll> special[k];
	for (ll i = 0; i < k; i++) {
		cin >> special[i].first >> special[i].second;
		--special[i].first;
		--special[i].second;
	}
	
	ll dists[k][n][m];
		
	for (ll i = 0; i < k; i++) {
		deque<pair<pair<ll, ll>, ll>> dq;
		dq.push_front(make_pair(special[i], grid[special[i].first][special[i].second]));
		
		bool vis[n][m];
		memset(vis, 0, sizeof(vis));
		
		while (!dq.empty()) {
			pair<pair<ll, ll>, ll> cur = dq.front();
			dq.pop_front();
			
			pair<ll, ll> loc = cur.first;
			ll r = loc.first, c = loc.second;
			ll dist = cur.second;
			
			if (vis[r][c]) continue;
			vis[r][c] = 1;
			
			dists[i][r][c] = dist;
			
			for (ll d = 0; d < 4; d++) {
				ll nr = r + dr[d], nc = c + dc[d];
				if (valid(nr, nc)) {
					if (grid[nr][nc]) {
						dq.push_back(make_pair(make_pair(nr, nc), dist + 1));
					} else {
						dq.push_front(make_pair(make_pair(nr, nc), dist));
					}
				}
			}
		}
	}
		
	ll edge_dist[n][m];
	
	deque<pair<pair<ll, ll>, ll>> dq;
	for (ll i = -1; i <= n; i++) {
		dq.push_back(make_pair(make_pair(i, -1), 0));
		dq.push_back(make_pair(make_pair(i, m), 0));
	}
	
	for (ll i = -1; i <= m; i++) {
		dq.push_back(make_pair(make_pair(-1, i), 0));
		dq.push_back(make_pair(make_pair(n, i), 0));
	}
		
	bool vis[n][m];
	memset(vis, 0, sizeof(vis));
		
	while (!dq.empty()) {
		pair<pair<ll, ll>, ll> cur = dq.front();
		dq.pop_front();
		
		pair<ll, ll> loc = cur.first;
		ll r = loc.first, c = loc.second;
		ll dist = cur.second;
		
		// cout << r << " " << c << " " << dist << '\n';
		
		if (valid(r, c)) {
			if (vis[r][c]) continue;
			vis[r][c] = 1;
			
			edge_dist[r][c] = dist;
		}
		
		for (ll d = 0; d < 4; d++) {
			ll nr = r + dr[d], nc = c + dc[d];
			if (valid(nr, nc)) {
				if (grid[nr][nc]) {
					dq.push_back(make_pair(make_pair(nr, nc), dist + 1));
				} else {
					dq.push_front(make_pair(make_pair(nr, nc), dist));
				}
			}
		}
	}
		
	ll best = 1e9;
	for (ll i = 0; i < k; i++) {
		for (ll j = i + 1; j < k; j++) {
			best = min(best, edge_dist[special[i].first][special[i].second] + edge_dist[special[j].first][special[j].second]);
		}
	}
		
	for (ll i = 0; i < n; i++) {
		for (ll j = 0; j < m; j++) {
			ll dist_vals[k];
			for (ll x = 0; x < k; x++) {
				dist_vals[x] = dists[x][i][j];
			}
			sort(dist_vals, dist_vals + k);
			
			ll cost = dist_vals[0] + dist_vals[1] + edge_dist[i][j];
			if (grid[i][j]) cost -= 2; // counted 3 times
			
			// cout << i << " " << j << " " << cost << " " << edge_dist[i][j] << '\n';
			
			best = min(best, cost);
		}
	}
		
	cout << best << '\n';
}
 
int main(int argc, char* argv[]) {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	int tc = 1;
	// cin >> tc;
	for (int t = 0; t < tc; t++) solve(argc, argv, t);
} 

File: archive/2020f/solutions/TheFlockofRams.java
import java.util.*;
import java.io.*;

// By Ayush Varshney
public class Rams {
	static int n, m, k;
	static int[] dr = new int[] { 0, 0, -1, 1 };
	static int[] dc = new int[] { 1, -1, 0, 0 };

	public static void main(String[] args) throws IOException {
		Scanner2 sc = new Scanner2(System.in);
		n = sc.nextInt();
		m = sc.nextInt();
		k = sc.nextInt();
		int[][] grid = new int[n][m];
		for (int i = 0; i < n; i++) {
			String s = sc.next();
			for (int j = 0; j < m; j++) {
				if (s.charAt(j) == '#') {
					grid[i][j] = 1;
				} else {
					grid[i][j] = 0;
				}
			}
		}
		int[][] special = new int[k][2];
		for (int i = 0; i < k; i++) {
			int a = sc.nextInt(), b = sc.nextInt();
			special[i] = new int[] { a - 1, b - 1 };
		}
		int[][][] dists = new int[k][n][m];
		for (int i = 0; i < k; i++) {
			ArrayDeque<int[][]> dq = new ArrayDeque<>();
			dq.addFirst(new int[][] { special[i], new int[] { grid[special[i][0]][special[i][1]] } });
			int[][] vis = new int[n][m];
			while (!dq.isEmpty()) {
				int[][] cur = dq.removeFirst();
				int[] loc = cur[0];
				int r = loc[0], c = loc[1];
				int dist = cur[1][0];
				if (vis[r][c] == 1)
					continue;
				vis[r][c] = 1;
				dists[i][r][c] = dist;
				for (int d = 0; d < 4; d++) {
					int nr = r + dr[d], nc = c + dc[d];
					if (valid(nr, nc)) {
						if (grid[nr][nc] == 1) {
							dq.addLast(new int[][] { new int[] { nr, nc }, new int[] { dist + 1 } });
						} else {
							dq.addFirst(new int[][] { new int[] { nr, nc }, new int[] { dist } });
						}
					}
				}
			}
		}
		int[][] edge_dist = new int[n][m];
		ArrayDeque<int[][]> dq = new ArrayDeque<>();
		for (int i = -1; i <= n; i++) {
			dq.addLast(new int[][] { new int[] { i, -1 }, new int[] { 0 } });
			dq.addLast(new int[][] { new int[] { i, m }, new int[] { 0 } });
		}
		for (int i = -1; i <= m; i++) {
			dq.addLast(new int[][] { new int[] { -1, i }, new int[] { 0 } });
			dq.addLast(new int[][] { new int[] { n, i }, new int[] { 0 } });
		}
		int[][] vis = new int[n][m];
		while (!dq.isEmpty()) {
			int[][] cur = dq.removeFirst();
			int[] loc = cur[0];
			int r = loc[0], c = loc[1];
			int dist = cur[1][0];
			if (valid(r, c)) {
				if (vis[r][c] == 1)
					continue;
				vis[r][c] = 1;
				edge_dist[r][c] = dist;
			}
			for (int d = 0; d < 4; d++) {
				int nr = r + dr[d], nc = c + dc[d];
				if (valid(nr, nc)) {
					if (grid[nr][nc] == 1) {
						dq.addLast(new int[][] { new int[] { nr, nc }, new int[] { dist + 1 } });
					} else {
						dq.addFirst(new int[][] { new int[] { nr, nc }, new int[] { dist } });
					}
				}
			}
		}
		int best = (int) 1e9;
		for (int i = 0; i < k; i++) {
			for (int j = i + 1; j < k; j++) {
				best = Math.min(best,
						edge_dist[special[i][0]][special[i][1]] + edge_dist[special[j][0]][special[j][1]]);
			}
		}
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				Integer[] dist_vals = new Integer[k];
				for (int x = 0; x < k; x++) {
					dist_vals[x] = dists[x][i][j];
				}
				Arrays.sort(dist_vals);
				int cost = dist_vals[0] + dist_vals[1] + edge_dist[i][j];
				if (grid[i][j] == 1)
					cost -= 2;
				best = Math.min(best, cost);
			}
		}
		System.out.println(best);
	}

	public static boolean valid(long r, long c) {
		return r >= 0 && r < n && c >= 0 && c < m;
	}

}

class Scanner2 {
	public BufferedReader reader;
	public StringTokenizer st;

	public Scanner2(InputStream stream) {
		reader = new BufferedReader(new InputStreamReader(stream));
		st = null;
	}

	public String next() {
		while (st == null || !st.hasMoreTokens()) {
			try {
				String line = reader.readLine();
				if (line == null)
					return null;
				st = new StringTokenizer(line);
			} catch (Exception e) {
				throw (new RuntimeException());
			}
		}
		return st.nextToken();
	}

	public int nextInt() {
		return Integer.parseInt(next());
	}

	public long nextLong() {
		return Long.parseLong(next());
	}

	public double nextDouble() {
		return Double.parseDouble(next());
	}
}